Allow preprocessors to specify extra C sources

ian-ross opened the pull request:
Alter return type from preprocessing to allow preprocessors like hsc2hs and C2HS to inform Cabal of extra C sources that they create that need to be compiled and linked.  I believe this fixes #238 as well as providing equivalent support for a related feature in new versions of C2HS where the preprocessor generates some C wrapper functions to help with argument marshalling.

23Skidoo commented:
This breaks the build on Travis.

Please add a changelog note. Maybe also a test and some documentation?

ian-ross commented:
@23Skidoo Will do.  I hadn't seen the warnings that are breaking the Travis build.  And where should documentation go?  Just inline comments, or is there a better place?  (Sorry if that's a dumb question: this is the first time I've contributed to Cabal...)

23Skidoo commented:
You probably compiled without `-Wall`. Documentation for user-facing features [lives under `Cabal/doc`](https://github.com/haskell/cabal/tree/master/Cabal/doc).

ian-ross commented:
@23Skidoo I think this is now better -- there's a test (that helped fix a problem...), a changelog comment, some documentation and the build on Travis is fine.  What's your preferred approach for merging PRs with extra commits?  Should I squash the commits into one and force push, close the PR and open another one with a squashed commit, or just leave it as is?

23Skidoo commented:
> Should I squash the commits into one and force push, close the PR and open another one with a squashed commit, or just leave it as is?

If you want to squash the commits, you can just do `push -f` after `rebase -i`. The PR will be updated. This sometimes makes code review comments attached to old diffs (in this case there aren't any) to vanish, though.

ian-ross commented:
OK, rebased and pushed.

23Skidoo commented on cabal-install/Distribution/Client/SetupWrapper.hs:
```diff
@@ -333,7 +333,9 @@ externalSetupMethod verbosity options pkg bt mkargs = do
     when srcNewer $ if useHs
                     then copyFileVerbose verbosity src setupHs
-                    else runSimplePreProcessor ppUnlit src setupHs verbosity
+                    else do
```
Isn't this do-block equivalent to `void $ runSimplePreProcessor ...`?

ian-ross replied:
Oops.  Yes.  I'll fix it.

23Skidoo commented on Cabal/Distribution/Simple/PreProcess.hs:
```diff
@@ -126,20 +126,29 @@ data PreProcessor = PreProcessor {
+  where inFile  = normalise (inBaseDir  </> inRelativeFile)
+        outFile = normalise (outBaseDir </> outRelativeFile)
+
+mkLessSimplePreProcessor :: (FilePath -> FilePath -> Verbosity -> IO [FilePath])
```
Not a very good name, IMO. Maybe add a comment explaining why it is "less simple".

ian-ross replied:
It's a *terrible* name.  It was supposed to be a placeholder while I thought of a better name...  I'll change it and add a comment.

23Skidoo commented:
LGTM with minor comments.

This changes the `PreProcessor` type, and by extension `PPSuffixHandler`, which is part of `UserHooks`. @dcoutts, are we OK with this?

23Skidoo commented:
So the main problem here is that this change breaks Setup scripts that define custom preprocessors. See [here](http://stackoverflow.com/questions/7370029/any-example-of-a-custom-preprocessor-in-haskell) for an example of such setup script.

ian-ross commented:
@23Skidoo OK, let me think of a way to do this in a way that's backwards compatible with the existing `PreProcessor` types.  I didn't really take on board that that stuff was user-visible.

23Skidoo commented:
This may be quite hard given that the `UserHooks` API is not really [designed for extensibility](http://www.yesodweb.com/blog/2015/03/designing-apis-for-extensibility). If we ever do a Cabal 2.0 release, we should take the opportunity to hide the internal structure of all user-visible records.

ian-ross commented:
Yeah, I was just looking at that.  It's going to be hard to do this, because the `PreProcessor` type is exported as `PreProcessor(..)` -- I changed `mkSimplePreProcessor` in a way that would work with that example you linked to, but if people can construct arbitrary `PreProcessor` instances, I can't see how to deal with a change in the type of `runPreProcessor` in a backwards compatible way.

23Skidoo commented:
There should at least be a way to write setup scripts that are compatible with both old and new versions of Cabal. It'd be also nice to check whether this change breaks any currently existing package on Hackage.

ian-ross commented:
That makes sense.  Let me collect some data: I'll look for all the packages on Hackage that use custom build scripts and see if any of them are broken by this change.  I'll also collect all the examples of custom preprocessors I can find.  That won't help with peoples' private code that's not on Hackage, but at least it's a start and should give some idea of how this is used in practice.  Results in a couple of days...

23Skidoo commented:
Cool, thank you for being so thorough.

ian-ross commented:
OK, this would break all the Gtk packages.  Don't see a way to get round it.

ian-ross commented:
@23Skidoo OK, I've now done this in a different way.  It's kind of ugly, but it doesn't change the user-visible preprocessor API at all, so it ought not to perturb any of the existing packages that use preprocessors in custom builds.  Given the constraints on changes to the public Cabal API, this seems like a better way to do things for now.

ian-ross commented:
@23Skidoo Thanks for the detailed review!  I've added Haddock comments, added full module names in the error messages, and have changed the way that the extra C sources are searched for, just looking for files matching the relevant pattern under the build directory (which is where these things end up, not in the autogen directory).

23Skidoo commented:
Thanks, everything looks good to me now. Merging.

ian-ross commented:
Great.  That was kind of epic.  Thanks for being patient.

23Skidoo commented:
Thanks to you for your contribution!
